lib/linked_list.rb

```diff
@@ -12,120 +12,404 @@ def initialize(value, next_node = nil)
 
 # Defines the singly linked list
 class LinkedList
-    def initialize
-      @head = nil # keep the head private. Not accessible outside this class
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
+  
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  def add_first(data)
+    current_first = @head
+    if @head == nil 
+      temp = Node.new(data)
+      @head = temp
+    else
+      temp = Node.new(data)
+      temp.next = @head
+      @head = temp
     end
-
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    def add_first(value)
-      raise NotImplementedError
+  end
+  
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  def search(value)
+    
+    current = @head
+    
+    if current == nil
+      return false
+    elsif current.data == nil
+      return false
     end
-
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    def search(value)
-      raise NotImplementedError
+    
+    until current.next == nil do

```

Comment:
What about if the last node is the one with `value`?  

```suggestion
    until current.nil? do
```

Then you don't need the if statement at the end.